pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,310 @@
+func (s *SSTSinkKeyWriter) WriteKey(ctx context.Context, key storage.MVCCKey, value []byte) error {
+	if len(s.flushedFiles) == 0 {
+		return errors.AssertionFailedf(
+			"no BackupManifest_File to write key to, call Reset before WriteKey",
+		)
+	}
+	if err := s.maybeFlushOOOKey(ctx, key.Key); err != nil {
+		return err
+	}
+	s.setMidRowForPrevKey(key.Key)
+	err := s.maybeDoSizeFlushWithNextKey(ctx, key.Key)
+	if err != nil {
+		return err
+	}
+
+	// At this point, because of the Reset invariant, we can make the following
+	// assumptions about the new key:
+	// - The new key comes after the previous key.
+	// - This new key belongs to the span of the last BackupManifest_File in
+	//   s.flushedFiles.
+	if len(s.flushedFiles) == 0 {
+		panic(
+			"unexpected missing BackupManifest_File after flushes when writing key",
+		)
+	}
+	elidedKey, prefix, err := elideMVCCKeyPrefix(key, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	s.elidedPrefix = append(s.elidedPrefix[:0], prefix...)
+	if err := s.sst.PutRawMVCC(elidedKey, value); err != nil {
+		return err
+	}
+
+	// TODO (kev-cao): Should add some method of tracking `Rows` (and potentially
+	// `IndexEntries`).
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(key.Key)) + int64(len(value)),
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	if lastFile.EntryCounts.DataSize >= fileSpanByteLimit {
+		// As we are splitting due to the fileSpanByteLimit, the new span will need
+		// to contiguously extend the last span. We can use the current key as an
+		// exclusive end key for the last file and an inclusive start key for the
+		// new file.
+		newSpan := roachpb.Span{
+			Key:    key.Key,
+			EndKey: lastFile.Span.EndKey,
+		}
+		lastFile.Span.EndKey = newSpan.Key
+		if err := s.Reset(ctx, newSpan); err != nil {
+			return err
+		}
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedSize += keyAsRowCount.DataSize
+	s.prevKey = append(s.prevKey[:0], key.Key...)
+
+	return nil
+}
+
+// Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
+// open BackupManifest_File exists for the new span.
+// Any time a new span is being written, Reset MUST be called prior to any
+// WriteKey calls.
+func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
+	if err := s.maybeFlushOOOSpans(ctx, newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	var isContiguous, isSame bool
+	if lastFile != nil {
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+		isSame = lastFile.Span.Equal(newSpan)
+		isContiguous = isContiguousSpan(lastFile.Span, newSpan)
+		if isSame {
+			return nil
+		} else if isContiguous &&
+			(s.midRow || lastFile.EntryCounts.DataSize < fileSpanByteLimit) {
```

Comment:
why are these conditionals here? We check `fileSpanByteLimit` elsewhere and `s.midRow` should never be true on reset.

Reply:
Let's say I write a span that ends `midRow` (which I'm not sure if that is protected against in `generateAndSplitSpans`), and then reset on a contiguous span afterwards. That's a case where you can have `isContiguous` be true and `s.midRow` be true. (See the `size-flush-postponed-till-after-mid-row` test case).

We could of course assert that this can't be possible, but I feel like that's also coupling the logic with how it is currently being used in backup compactions.

---

> We check `fileSpanByteLimit` elsewhere

In the case where we check this, that's only if you are writing keys that belong to the same span. If you write keys that belong to a different contiguous span, then you call `Reset()` with that new span.

Reply:
Oh just came back to this a bit later and realized what you meant. I think the only case for checking `fileSpanByteLimit` then is really just to determine if we should increment `spanGrows`.

Reply:
> Let's say I write a span that ends midRow (which I'm not sure if that is protected against in generateAndSplitSpans)

This will never happen, and i don't think it is the sink's responsibility to check this.

Reply:
> then is really just to determine if we should increment spanGrows.
It doesn't seem like we use these stats anywhere. if managing them significantly complicates the code, i'm cool with ignoring them for now.